### DISCUSSION TITLE
[C128] Maximize the available memory for execution and pointers into other memory banks

### DISCUSSION THREAD

**Discussion by RetrofanC** [cc65/cc65#1567] [Q&A] [answered]
Hi all, 

I have been using the cc65 tool-set on different occasions and really appreciate it's existence the people keeping it alive!
 
Currently I am working on a program for the C128 which demands a bit more than the default profiles and settings. So this time I am looking into a way the maximize the amount of memory available for running code. Now I have been looking to the different documentation and forums and the items which came most close was the closed issue #772 so want to build a bit on that.

What I understand is the following but that also raises the question I put in between; 

When I look into the C128.cfg and read the cc65 C128 specific configuration it defines; 
c128.cfg: start = $1C0D and size $A3F3 so ends at $C000.  
cc65 doc: start = $1C00 till $BFFF where the stack builds down. 

****[Q1]** Why does the memory stops at $C000 and isn't defined until $D000?** 

Now when I use overlays in the c128-overlay.cfg it defines the overlays as;
start = $C000 which is outside the regular range which is till $C000.

****[Q2]** Does this means I can just change the value in c128.cfg for size from $A3F3 into ($A3F3 + $1000)? 
If I never intend to use the overlay functionality?** 

****[Q3]** Alternatively, when I do want to use the overlay functionality but I want to use an overlay size bigger than 4k lets say 8k. Should I then lower the (overlay) address start from $C000 to ($C000 - $1000) and reduce the $A3F3 to ($A3F3 - $1000)?** 

Or... when I take it one step further

****[Q4]** Can I swap out the I/O and Kernel for RAM and make the size ($A3F3 + $1000+ $1000(RAM under I/O) +$2000 (RAM under kernel))?** 
_(Being fully conscious that I have to swap in the I/O myself whenever I needed it in my code. Furthermore, I have to write the needed kernel routines myself. Which both are not problem.)_

Now if I still want to keep the I/O in place but are fully happy to disable the kernel...
****[Q5]** Is there any way in which I can make the c128.cfg (or different files of the target definition) to make it use this total memory available despite there is a 4k gap in the memory because of the $D000 I/O space?** 

If the answer to the above would be "No" or very difficult.... 
****[Q6]** Would I alternatively be able to just define the end of memory to $D000 in that case leaving the I/O intact and define the overlay space starting at $E000 for a max of 8k (minus few bytes so only until $FF00 I guess).** 

To conclude a related question but from a very different angle. However, it could be considered less cc65 relevant but more programing technics. However, depending on the answer it could become and addition to the cc65 tool-set. 

Within my program I use several arrays as well and instead of arrays I could use pointers to memory locations of course. Would there be a relative simple way (code examples) in which I can define an array (or pointer) read/write in such a way that it refers to the memory of a different memory bank but doesn't interrupts the program execution by clever bank switching? I think somewhere in the lines of how the kernel routines INDSTA & INDFET are working...

That's all for now, thanks in advance for your answers and guidance on the above! 

PS Surely any insights or ideation on the above not being c128 specific but from other targets are welcome as well of course.

**Comment by oliverschmidt** [comment]
> Now when I use overlays in the c128-overlay.cfg it defines the overlays as;
start = $C000 which is outside the regular range which is till $C000.

I don't see 'start = $C000' anywhere in https://github.com/cc65/cc65/blob/master/cfg/c128-overlay.cfg. As far as I can see, the overlays start (by default) at $B000, see https://github.com/cc65/cc65/blob/master/cfg/c128-overlay.cfg#L12.

> Should I then lower the (overlay) address start from $C000 to ($C000 - $1000) and reduce the $A3F3 to ($A3F3 - $1000)?

No, https://github.com/cc65/cc65/blob/master/cfg/c128-overlay.cfg computes the start address from the size, see e.g. https://github.com/cc65/cc65/blob/master/cfg/c128-overlay.cfg#L13

**Reply by RetrofanC** [reply]
Thanks for the quick reply and perhaps I did read the c128-ovelay.cfg incorrect.

Let me follow your logic here; 

#L12:   MAIN:     file = %O, define = yes, start = $1C0D,                   size = $A3F3 - __OVERLAYSIZE__ - __STACKSIZE__;

If start = $1C0D and the (default) size = $A3F3 that means end of (default) memory is  $1C0D + $A3F3 = $C000 correct? 
Then first the OVERLAYSIZE_(default $1000) is subtracted which would make it $C000 - $1000 = $B000.
(Which might the $B000 you are referring to is that correct?) 
But in reality only the actual overlay size is subtracted which is determined by the linker based on the size of the overlay file correct?

The actual start address of the overlay is put into $BFFE and $BFFF hence the size = $0002?

If my adjusted understanding above is correct my confusion came from this line; ; 
#L14:    OVL1:     file = "%O.1",                **start = $C000** - __OVERLAYSIZE__, size = __OVERLAYSIZE__;

Which I read as start = $C000 but should be $C000 subtracted with the OVERLAYSIZE is the actual start instead then. 

If above is correct it surely helps the understanding. 

What about expanding the size from $A3F3 to $B3F3 and possible beyond by disabling I/O and Kernel? 
Can that just be done and tried out or does something breaks in the tool-chain when I do that?

**Comment by oliverschmidt** [comment]
> [...] Which might the $B000 you are referring to is that correct?

Yes. 

> But in reality only the actual overlay size is subtracted which is determined by the linker based on the size of the overlay file correct?

No, what makes you think so? And in the first place, there's no THE actual overlay size as there are several overlays. So there would need to be code to calculate the maximum of all actual overlay sizes - but there isn't anywhere.

> What about expanding the size from $A3F3 to $B3F3 [...]

I have to leave it to somebody else to comment on that question.

**Reply by RetrofanC** [reply]
[OS] ....No, what makes you think so? And in the first place, there's no THE actual overlay size as there are several overlays. So there would need to be code to calculate the maximum of all actual overlay sizes - but there isn't anywhere... 

Well my reasoning was that when I look into the .map file of the compiled "overlaydemo" I see this; 
.....
    OVERLAY1          Offs=000000  Size=000044  Align=00001  Fill=0000
    OVERLAY2          Offs=000000  Size=000055  Align=00001  Fill=0000
    OVERLAY3          Offs=000000  Size=000032  Align=00001  Fill=0000
....
Hence I was assuming that if I combine the above with the earlier line #14 from the c128-overlay.cfg file (mentioned in the earlier msg) it would result into something like; 

  #L14: OVL1: file = "%O.1", start = $C000 - $44,  size = $44 ;

However, looking further in the .map file that isn't what actually is happening it seems. 
So what I understand currently is this; 

It takes the defined __OVERLAYSIZE__ from here; 
  __OVERLAYSIZE__: type = weak, value = $1000; # 4k overlay

Uses that in here to come to a new value for the 'start' OVL1ADDRESS and OVL1;
    OVL1ADDR: file = "%O.1",           start = $BFFE - __OVERLAYSIZE__, size = $0002;
    OVL1:     file = "%O.1",           start = $C000 - __OVERLAYSIZE__, size = __OVERLAYSIZE__;

Which becomes; 
OVL1ADDR start =$AFFE 
OVL1          start =$B000
(which seem to match the mentioned values in the .map) 

Then the actual compiled size of the 'ovrlay1' it is taken as the size and memory range for that 'ovrlay1'
As shown a bit further in the .map file. 
   OVERLAY1              00B000  00B043  000044  00001

   OVERLAY2              00B000  00B054  000055  00001
   OVERLAY3              00B000  00B031  000032  00001

Am I getting closer to how it actually works indeed? 

For the increase of the $A3F3 value I will just trail and error increasing that value and see what happens. 
Furthermore, I noticed groepaz answer to the the question [..configure ld65 to include a gap..]
I will look into that as well as that might give guidance for my question on freeing up kernal RAM but leave I/O in place.

Thanks again sofar, all support is appreciated!

**Comment by oliverschmidt** [comment]
> [...] Am I getting closer to how it actually works indeed?

As far as I see, what you describe matches my picture :-)

**Reply by RetrofanC** [reply]
Thanks all good now, I have enough information to take it to the next level. 

Few last related questions and then you can close this threat as answered. 
(If I need to open a new threat for the questions below please let me know as well and I will do that) 

I have done some further research and to change the memory configuration at start-up I need to change the c128.inc and crt0.s 
In c128.inc I will add an new MMU_CFG_CC65_CFG_2

Consequentially, I will change the crt0.s into; 


        lda     MMU_CR                             ; Get current memory configuration...
        pha                                                ; ...and save it for later
      **lda     #MMU_CFG_CC65_CFG_2   ; New C128 Memory Config **
         sta     MMU_CR

Just to make sure; 

[Q1] Is my understanding correct that in the cc65 toolchain the crt0.s is part of the target library so in my case c128.lib? 

[Q2] Is my understanding correct that if I want to use a changed crt0.s I need add it to and recompile a new c128.lib?

[Q3] Is there any quick way to ' take out" the existing crt0.o from the c128.lib and replace it with the new 'crt0.o'?
        If this is possible what are the commands to do this? 

[Q4] If the answer on Q3 is no, it is not possible, what are all the files that are compiled into the c128.lib?   
        Or is there a command which 'unwraps' all .o files contained in the c128.lib? 

Thanks! 
Getting there....

**Reply by polluks** [reply]
You may also take a look at Extras
https://cc65.github.io/doc/c64.html#ss5.1

**Comment by oliverschmidt** [comment]
> [Q1] Is my understanding correct that in the cc65 toolchain the crt0.s is part of the target library so in my case c128.lib?

Yes.

> [Q2] Is my understanding correct that if I want to use a changed crt0.s I need add it to and recompile a new c128.lib?

No.

Background: Any module (aka .o file) inside a library is only considered by the linker if the linker needs to resolve symbols. So if you make sure that the symbols provided by crt0.o are already present by the time the linker resolves symbols, the linker will happily ignore the crt0.o inside the library.

Howto: provide your crt0.o on the linker command line __BEFORE__ the c128.lib.

> [Q3] Is there any quick way to ' take out" the existing crt0.o from the c128.lib and replace it with the new 'crt0.o'?
If this is possible what are the commands to do this?

Yes. See https://cc65.github.io/doc/ar65.html

**Reply by RetrofanC** [reply]
Thanks! (-8  

Makes all sense and I know enough now to try to make the max memory c128. 

When I got it all working I will provide the results back here. 
This topic can be closed!

**Reply by RetrofanC** [reply]
Hi, 

As this thread is still open I will add my related question here. 

The stretching of memory works ok for the moment I have changed; 

$A3F3 -> $B3F3 for the available memory size making it till $D000 
$1000 -> $2000 for the size of the overlay and make go back from $D000 instead of the original $C000

Now I have a need as well to increase the number of overlays from 9 -> 12 so made 3 extra overlays and numbered them X,Y,Z. 

Consequently my c128-overlay.cfg looks like this; 

SYMBOLS {
    __LOADADDR__:    type = import;
    __EXEHDR__:      type = import;
    __OVERLAYADDR__: type = import;
    __STACKSIZE__:   type = weak, value = $0800; # 2k stack
    __OVERLAYSIZE__: type = weak, value = $2000; # 8k overlay
}
MEMORY {
    ZP:       file = "", define = yes, start = $0002,                   size = $001A;
    LOADADDR: file = %O,               start = $1BFF,                   size = $0002;
    HEADER:   file = %O,               start = $1C01,                   size = $000C;
    MAIN:     file = %O, define = yes, start = $1C0D,                   size = $B3F3 - __OVERLAYSIZE__ - __STACKSIZE__;
    OVL1ADDR: file = "%O.1",           start = $BFFE - __OVERLAYSIZE__, size = $0002;
    OVL1:     file = "%O.1",           start = $D000 - __OVERLAYSIZE__, size = __OVERLAYSIZE__;
    OVL2ADDR: file = "%O.2",           start = $BFFE - __OVERLAYSIZE__, size = $0002;
    OVL2:     file = "%O.2",           start = $D000 - __OVERLAYSIZE__, size = __OVERLAYSIZE__;
    OVL3ADDR: file = "%O.3",           start = $BFFE - __OVERLAYSIZE__, size = $0002;
    OVL3:     file = "%O.3",           start = $D000 - __OVERLAYSIZE__, size = __OVERLAYSIZE__;
    OVL4ADDR: file = "%O.4",           start = $BFFE - __OVERLAYSIZE__, size = $0002;
    OVL4:     file = "%O.4",           start = $D000 - __OVERLAYSIZE__, size = __OVERLAYSIZE__;
    OVL5ADDR: file = "%O.5",           start = $BFFE - __OVERLAYSIZE__, size = $0002;
    OVL5:     file = "%O.5",           start = $D000 - __OVERLAYSIZE__, size = __OVERLAYSIZE__;
    OVL6ADDR: file = "%O.6",           start = $BFFE - __OVERLAYSIZE__, size = $0002;
    OVL6:     file = "%O.6",           start = $D000 - __OVERLAYSIZE__, size = __OVERLAYSIZE__;
    OVL7ADDR: file = "%O.7",           start = $BFFE - __OVERLAYSIZE__, size = $0002;
    OVL7:     file = "%O.7",           start = $D000 - __OVERLAYSIZE__, size = __OVERLAYSIZE__;
    OVL8ADDR: file = "%O.8",           start = $BFFE - __OVERLAYSIZE__, size = $0002;
    OVL8:     file = "%O.8",           start = $D000 - __OVERLAYSIZE__, size = __OVERLAYSIZE__;
    OVL9ADDR: file = "%O.9",           start = $BFFE - __OVERLAYSIZE__, size = $0002;
    OVL9:     file = "%O.9",           start = $D000 - __OVERLAYSIZE__, size = __OVERLAYSIZE__;
    OVLXADDR: file = "%O.X",           start = $BFFE - __OVERLAYSIZE__, size = $0002;
    OVLX:     file = "%O.X",           start = $D000 - __OVERLAYSIZE__, size = __OVERLAYSIZE__;
    OVLYADDR: file = "%O.Y",           start = $BFFE - __OVERLAYSIZE__, size = $0002;
    OVLY:     file = "%O.Y",           start = $D000 - __OVERLAYSIZE__, size = __OVERLAYSIZE__;
    OVLZADDR: file = "%O.Z",           start = $BFFE - __OVERLAYSIZE__, size = $0002;
    OVLZ:     file = "%O.Z",           start = $D000 - __OVERLAYSIZE__, size = __OVERLAYSIZE__;


}
SEGMENTS {
    ZEROPAGE: load = ZP,       type = zp;
    LOADADDR: load = LOADADDR, type = ro;
    EXEHDR:   load = HEADER,   type = ro;
    STARTUP:  load = MAIN,     type = ro;
    LOWCODE:  load = MAIN,     type = ro,                optional = yes;
    ONCE:     load = MAIN,     type = ro,                optional = yes;
    CODE:     load = MAIN,     type = ro;
    RODATA:   load = MAIN,     type = ro;
    DATA:     load = MAIN,     type = rw;
    INIT:     load = MAIN,     type = bss;
    BSS:      load = MAIN,     type = bss, define = yes;
    OVL1ADDR: load = OVL1ADDR, type = ro;
    OVERLAY1: load = OVL1,     type = ro,  define = yes, optional = yes;
    OVL2ADDR: load = OVL2ADDR, type = ro;
    OVERLAY2: load = OVL2,     type = ro,  define = yes, optional = yes;
    OVL3ADDR: load = OVL3ADDR, type = ro;
    OVERLAY3: load = OVL3,     type = ro,  define = yes, optional = yes;
    OVL4ADDR: load = OVL4ADDR, type = ro;
    OVERLAY4: load = OVL4,     type = ro,  define = yes, optional = yes;
    OVL5ADDR: load = OVL5ADDR, type = ro;
    OVERLAY5: load = OVL5,     type = ro,  define = yes, optional = yes;
    OVL6ADDR: load = OVL6ADDR, type = ro;
    OVERLAY6: load = OVL6,     type = ro,  define = yes, optional = yes;
    OVL7ADDR: load = OVL7ADDR, type = ro;
    OVERLAY7: load = OVL7,     type = ro,  define = yes, optional = yes;
    OVL8ADDR: load = OVL8ADDR, type = ro;
    OVERLAY8: load = OVL8,     type = ro,  define = yes, optional = yes;
    OVL9ADDR: load = OVL9ADDR, type = ro;
    OVERLAY9: load = OVL9,     type = ro,  define = yes, optional = yes;
    OVLXADDR: load = OVLXADDR, type = ro;
    OVERLAYX: load = OVLX,     type = ro,  define = yes, optional = yes;
    OVLYADDR: load = OVLYADDR, type = ro;
    OVERLAYY: load = OVLY,     type = ro,  define = yes, optional = yes;
    OVLZADDR: load = OVLZADDR, type = ro;
    OVERLAYZ: load = OVLZ,     type = ro,  define = yes, optional = yes;


}
FEATURES {
    CONDES: type    = constructor,
            label   = __CONSTRUCTOR_TABLE__,
            count   = __CONSTRUCTOR_COUNT__,
            segment = ONCE;
    CONDES: type    = destructor,
            label   = __DESTRUCTOR_TABLE__,
            count   = __DESTRUCTOR_COUNT__,
            segment = RODATA;
    CONDES: type    = interruptor,
            label   = __INTERRUPTOR_TABLE__,
            count   = __INTERRUPTOR_COUNT__,
            segment = RODATA,
            import  = __CALLIRQ__;
}


It all builds/compiles nicely and the .map file shows the additional overlays as well; 

multidemo.o:
    CODE              Offs=000000  Size=0005B5  Align=00001  Fill=0000
    RODATA            Offs=000000  Size=000351  Align=00001  Fill=0000
    BSS               Offs=000000  Size=005B29  Align=00001  Fill=0000
    DATA              Offs=000000  Size=00001A  Align=00001  Fill=0000
    OVERLAY1          Offs=000000  Size=001F5F  Align=00001  Fill=0000
    OVERLAY2          Offs=000000  Size=001C5B  Align=00001  Fill=0000
    OVERLAY3          Offs=000000  Size=0019DB  Align=00001  Fill=0000
    OVERLAY4          Offs=000000  Size=00142D  Align=00001  Fill=0000
    OVERLAY5          Offs=000000  Size=001FC2  Align=00001  Fill=0000
    OVERLAY6          Offs=000000  Size=001DDF  Align=00001  Fill=0000
    OVERLAY7          Offs=000000  Size=001ACD  Align=00001  Fill=0000
    OVERLAY8          Offs=000000  Size=001FF0  Align=00001  Fill=0000
    OVERLAY9          Offs=000000  Size=001FF0  Align=00001  Fill=0000
    OVERLAYX          Offs=000000  Size=001B5C  Align=00001  Fill=0000
    OVERLAYY          Offs=000000  Size=00153E  Align=00001  Fill=0000
    OVERLAYZ          Offs=000000  Size=000B2B  Align=00001  Fill=0000

=======================================================

However, I do get these three warnings during the build;  

![image](https://user-images.githubusercontent.com/66948775/136204366-dda5ed75-00af-4181-8d89-bfe099be724e.png)

**[Q1]** What do I need to do to get ride of these warnings? 
              Where (in which file) do I need to make changes to satisfy the toolchain that all is ok with these extra added overlays...? 

Thanks!

**Comment by oliverschmidt** [comment, marked as answer]
> Where (in which file) do I need to make changes to satisfy the toolchain that all is ok with these extra added overlays...?

https://github.com/cc65/cc65/blob/master/libsrc/cbm/overlayaddr.s

**Reply by RetrofanC** [reply]
Brilliant! Thanks a lot (-8
